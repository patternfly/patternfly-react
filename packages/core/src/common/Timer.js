/**
 * Timer class implements a simple timeout mechanism
 */
class Timer {
  constructor(func, delay) {
    this.timer = null;
    this.delay = delay;
    this.execute = func;
  }

  // startTimer optionally takes a new func and delay so the timer instance can be reused.
  startTimer(func, delay = -1) {
    this.clearTimer();
    if (func) this.execute = func;
    if (delay >= 0) this.delay = delay;
    this.timer = setTimeout(this.execute, this.delay);
  }
  clearTimer() {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }
  skipTimer() {
    if (this.timer) {
      this.execute(true); // execute can take an optional `skipped` argument
      this.clearTimer();
    }
  }
}

export default Timer;
