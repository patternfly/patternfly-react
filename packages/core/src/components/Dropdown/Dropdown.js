export { Dropdown as default } from 'react-bootstrap';
