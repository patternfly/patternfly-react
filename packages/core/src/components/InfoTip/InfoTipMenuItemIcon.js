import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Icon } from '../Icon';

const InfoTipMenuItemIcon = ({ className, ...props }) => {
  const infoTipMenuItemIconClass = classNames('i', className);

  return <Icon className={infoTipMenuItemIconClass} {...props} />;
};

InfoTipMenuItemIcon.propTypes = {
  type: PropTypes.oneOf(['fa', 'pf']),
  name: PropTypes.string,
  className: PropTypes.string
};

InfoTipMenuItemIcon.defaultProps = {
  type: 'pf',
  name: 'info',
  className: ''
};

export default InfoTipMenuItemIcon;
