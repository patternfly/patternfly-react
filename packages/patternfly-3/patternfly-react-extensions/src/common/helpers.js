import React from 'react';

/** Implementation of the debounce function */
export const debounce = (func, wait) => {
  let timeout;
  function innerFunc(...args) {
    const context = this;
    clearTimeout(timeout);
    timeout = setTimeout(() => func.apply(context, args), wait);
  }
  return innerFunc;
};

/** Returns a subset of the given object including only the given keys, with values optionally replaced by a fn. */
export const selectKeys = (obj, keys, fn = val => val) =>
  keys.reduce((values, key) => ({ ...values, [key]: fn(obj[key]) }), {});

/** Returns a subset of the given object with a validator function applied to its keys. */
export const filterKeys = (obj, validator) => selectKeys(obj, Object.keys(obj).filter(validator));

/** Returns a subset of the given object with the given keys left out. */
export const excludeKeys = (obj, keys) => filterKeys(obj, key => !keys.includes(key));

/** Returns the given React children prop as a regular array of React nodes. */
export const childrenToArray = children =>
  children && React.Children.count(children) > 0 && React.Children.toArray(children);

/** Returns true if the component has the desired displayName value */
export const hasDisplayName = (component, displayName) =>
  component && component.type && component.type.displayName === displayName;

export const noop = Function.prototype;

export const helpers = {
  debounce,
  selectKeys,
  filterKeys,
  excludeKeys,
  childrenToArray,
  hasDisplayName,
  noop
};

export default helpers;
