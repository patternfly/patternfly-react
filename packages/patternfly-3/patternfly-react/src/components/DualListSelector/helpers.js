export const filterByHiding = (list, value) => {
  const filterValue = value.toLowerCase();
  return list.map(item => {
    const itemLabel = item.label.toLowerCase();
    const included = itemLabel.includes(filterValue);
    // if the item label matches the filter value.
    item.hiddenByFilter = !included;
    // if it is a parent and its label doesn't match the filter value.
    if (itemHasChildren(item)) {
      if (isItemHiddenByFilter(item)) {
        let childrenIncludedAmount = 0;
        item.children.forEach(childItem => {
          if (childItem.hidden) {
            return;
          }
          const childLabel = childItem.label.toLowerCase();
          const childIncluded = childLabel.includes(filterValue);
          childItem.hiddenByFilter = !childIncluded;
          childrenIncludedAmount += childIncluded ? 1 : 0;
        });
        item.hiddenByFilter = childrenIncludedAmount === 0;
      } else {
        item.children = makeAllHiddenFilteredItemsVisible(item.children);
      }
    }
    return item;
  });
};

// We don't want to touch default hidden items that the consumer set to hidden.
export const makeAllHiddenFilteredItemsVisible = list =>
  list.map(item => {
    item.hiddenByFilter = false;
    if (itemHasChildren(item)) {
      item.children.forEach(childItem => {
        childItem.hiddenByFilter = false;
      });
    }
    return item;
  });

export const sortItems = (items, sortFactor = 'label') =>
  items.sort((a, b) => (a[sortFactor].toLowerCase() > b[sortFactor].toLowerCase() ? 1 : -1));

export const shouldItemBeChecked = (item, isMainChecked, resetAllSelected) => {
  let checked = item.checked || false;
  const isItemEditable = !item.disabled || !item.hidden || !item.isItemHiddenByFilter;
  if (!isItemEditable) {
    return checked;
  }
  if (resetAllSelected) {
    checked = false;
  } else if (isMainChecked) {
    checked = isMainChecked;
  }
  return checked;
};

export const arrangeArray = ({ items, sortBy, isSortAsc = true, isMainChecked = false, resetAllSelected = false }) => {
  // sort the items
  let itemsCopy = sortItems(items, sortBy).map((item, index) => {
    // add position to the item and update if the main checkbox is initialy checked.
    const modifiedItem = {
      ...item,
      position: index,
      checked: shouldItemBeChecked(item, isMainChecked, resetAllSelected)
    };
    if (itemHasChildren(item)) {
      // sort the children array and add a position, parentPosition and update check state.
      modifiedItem.children = sortItems(item.children, sortBy).map((child, childIndex) => ({
        ...child,
        position: childIndex,
        parentPosition: index,
        checked: shouldItemBeChecked(child, isMainChecked, resetAllSelected)
      }));
    }
    return modifiedItem;
  });

  itemsCopy = isSortAsc ? itemsCopy : reverseAllItemsOrder(itemsCopy);

  return itemsCopy;
};

export const getDefaultProps = () => ({
  items: [],
  options: null,
  isSortAsc: true,
  sortBy: 'label',
  filterTerm: '',
  isMainChecked: false
});

export const getCheckedAmount = items => {
  let checkedAmount = 0;
  items.forEach(item => {
    if (isItemSelected(item)) {
      checkedAmount += 1;
      if (itemHasChildren(item)) {
        checkedAmount += item.children.length;
      }
    } else if (itemHasChildren(item)) {
      item.children.forEach(child => {
        if (isItemSelected(child)) {
          checkedAmount += 1;
        }
      });
    }
  });
  return checkedAmount;
};
export const getCounterMessage = (selected, total) => `${selected} of ${total} items selected`;

export const adjustProps = ({ left, right, ...props }) => {
  const defaultProps = getDefaultProps();
  const leftItems = arrangeArray({ ...left });
  const rightItems = arrangeArray({ ...right });
  return {
    ...props,
    left: {
      ...defaultProps,
      ...left,
      items: leftItems,
      selectCount: getCheckedAmount(leftItems)
    },
    right: {
      ...defaultProps,
      ...right,
      items: rightItems,
      selectCount: getCheckedAmount(rightItems)
    }
  };
};

export const isAllChildrenChecked = ({ children }) =>
  children && children.filter(({ checked }) => checked).length === children.length;

export const getItemsLength = items => {
  let { length } = items;
  let hiddenItemsAmount = 0;
  if (length === 0) {
    return 0;
  }
  items.forEach(({ hidden, children }) => {
    if (hidden) {
      hiddenItemsAmount += 1;
    }
    if (children) {
      length -= 1;
      children.forEach(child => {
        if (!child.hidden) {
          length += 1;
        }
      });
    }
  });
  return length - hiddenItemsAmount;
};

export const reverseAllItemsOrder = items => {
  const reversedItems = [...items].reverse();
  return reversedItems.map(item => (item.children ? { ...item, children: item.children.reverse() } : item));
};

export const getItem = (isSortAsc, items, position, parentPosition) => {
  // if item is a child.
  if (parentPosition !== undefined) {
    const parent = items[getItemPosition(items, parentPosition, isSortAsc)];
    return parent.children[getItemPosition(parent.children, position, isSortAsc)];
  }
  return items[getItemPosition(items, position, isSortAsc)];
};

export const getUpdatedSelectCount = (selectCount, checked, amount = 1) =>
  selectCount + (checked ? amount : -1 * amount);

export const itemHasParent = item => item.parentPosition !== undefined;

export const itemHasChildren = item => item.children !== undefined;

export const getItemPosition = (array, position, isSortAsc) => (isSortAsc ? position : array.length - position - 1);

export const toggleAllItems = (list, checked) => {
  let toggleCount = 0;
  list.forEach(item => {
    if (item.disabled || item.hidden) {
      return;
    }
    if (item.checked !== checked) {
      item.checked = checked;
      toggleCount += 1;
    }
    if (itemHasChildren(item)) {
      let childrenToggleCount = 0;
      item.children.forEach(childItem => {
        if (childItem.disabled || childItem.hidden) {
          return;
        }
        if (childItem.checked !== checked) {
          childItem.checked = checked;
          childrenToggleCount += 1;
        }
      });
      if (childrenToggleCount > 0) {
        toggleCount += childrenToggleCount - 1;
      }
    }
  });
  return toggleCount;
};

export const isAllItemsChecked = (items, selectCount) => selectCount > 0 && selectCount === getItemsLength(items);

export const isItemExistOnList = (list, itemLabel) => {
  let parentIndex = null;
  // find if the parent already exist on the list.
  list.forEach((listItem, index) => {
    if (listItem.label === itemLabel) {
      parentIndex = index;
    }
  });
  return { isParentExist: parentIndex !== null, parentIndex };
};

export const getFilterredItems = list => {
  const filteredItems = [];
  list.forEach(item => {
    if (!isItemHiddenByFilter(item)) {
      filteredItems.push(item);
    } else if (itemHasChildren(item)) {
      const filteredChildren = [];
      item.children.forEach(childItem => {
        if (!isItemHiddenByFilter(childItem)) {
          filteredChildren.push(childItem);
        }
      });
      if (filteredChildren.length > 0) {
        filteredItems.push({ ...item, children: filteredChildren });
      }
    }
  });

  return filteredItems;
};

export const getFilterredItemsLength = list => getItemsLength(getFilterredItems(list));

export const getSelectedFilterredItemsLength = list => {
  const filteredItems = getFilterredItems(list);
  let selectedAmount = 0;
  filteredItems.forEach(item => {
    if (isItemSelected(item)) {
      selectedAmount += 1;
      if (itemHasChildren(item)) {
        let selectedChildrenAmount = 0;
        item.children.forEach(childItem => {
          if (isItemSelected(childItem)) {
            selectedChildrenAmount += 1;
          }
        });
        if (selectedChildrenAmount) {
          selectedAmount += selectedChildrenAmount - 1;
        }
      }
    }
  });
  return selectedAmount;
};

export const isItemSelected = item => item.checked;

export const isItemHiddenByFilter = item => item.hiddenByFilter;

export const isItemDisabled = item => item.disabled;
