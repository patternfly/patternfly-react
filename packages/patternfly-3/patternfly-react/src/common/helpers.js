import React from 'react';
import { default as Timer } from './Timer';
import { closest as closestPolyfill } from './closestPolyfill';
import { default as controlled } from './controlled';
import { patternfly, c3ChartDefaults, layout } from './patternfly';

/** Equivalent to calling `this.someMethod = this.someMethod.bind(this)` for every method name in the methods array. */
export const bindMethods = (context, methods) => {
  // eslint-disable-next-line no-console
  console.warn(`
   bindMethods usage is deprecated in favor of class methods.
   bindMethods will be removed in the next major release
   `);
  methods.forEach(method => {
    context[method] = context[method].bind(context);
  });
};

/** Implementation of the debounce function */
export const debounce = (func, wait) => {
  let timeout;
  function innerFunc(...args) {
    const context = this;
    clearTimeout(timeout);
    timeout = setTimeout(() => func.apply(context, args), wait);
  }
  return innerFunc;
};

/** Returns true if propName is a non-null, defined property of the props object (can be any object, not just React props). */
export const propExists = (props, propName) => props && props.hasOwnProperty(propName) && props[propName] != null;

/** Given two objects (props and state), returns the value of propName from props if present, or from state otherwise. */
export const propOrState = (props, state, propName) =>
  propExists(props, propName) ? props[propName] : state[propName];

/** Returns a subset of the given object including only the given keys, with values optionally replaced by a fn. */
export const selectKeys = (obj, keys, fn = val => val) =>
  keys.reduce((values, key) => ({ ...values, [key]: fn(obj[key]) }), {});

/** Returns a subset of the given object with a validator function applied to its keys. */
export const filterKeys = (obj, validator) => selectKeys(obj, Object.keys(obj).filter(validator));

/** Returns a subset of the given object with the given keys left out. */
export const excludeKeys = (obj, keys) => filterKeys(obj, key => !keys.includes(key));

/** Returns the given React children prop as a regular array of React nodes. */
export const childrenToArray = children =>
  children && React.Children.count(children) > 0 && React.Children.toArray(children);

/** Filters the given React children prop with the given validator function. Returns an array of nodes. */
export const filterChildren = (children, validator) => {
  const array = childrenToArray(children);
  return array && array.filter(validator);
};

/** Given a React children prop, finds the first child node to pass the validator function. */
export const findChild = (children, validator) => {
  const array = childrenToArray(children);
  return array && array.find(validator);
};

/** Returns true if there is at least one of propNames with a different value in newProps than in oldProps. */
export const propsChanged = (propNames, oldProps, newProps) =>
  propNames.some(propName => oldProps[propName] !== newProps[propName]);

/** Returns true if the component has the desired displayName value */
export const hasDisplayName = (component, displayName) =>
  component && component.type && component.type.displayName === displayName;

/** Returns an object with the same keys as the given one, but all null values. */
export const nullValues = obj => selectKeys(obj, Object.keys(obj), () => null);

export const noop = Function.prototype;

export const KEY_CODES = {
  TAB_KEY: 9,
  ENTER_KEY: 13,
  ESCAPE_KEY: 27,
  SHIFT: 16,
  A: 65,
  Z: 90,
  NUMPAD: { 0: 97 },
  F11: 122
};
export const KEYS = {
  ENTER: 'Enter',
  CAPSLOCK: 'CapsLock',
  SPACE: ' ',
  ARROW_LEFT: 'ArrowLeft',
  ARROW_RIGHT: 'ArrowRight',
  ARROW_DOWN: 'ArrowDown',
  ARROW_UP: 'ArrowUp',
  HOME: 'Home',
  END: 'End'
};

export const helpers = {
  Timer,
  closestPolyfill,
  controlled,
  patternfly,
  c3ChartDefaults,
  layout,
  debounce,
  propExists,
  propOrState,
  selectKeys,
  filterKeys,
  excludeKeys,
  childrenToArray,
  filterChildren,
  findChild,
  propsChanged,
  hasDisplayName,
  nullValues,
  noop,
  KEY_CODES,
  KEYS
};
