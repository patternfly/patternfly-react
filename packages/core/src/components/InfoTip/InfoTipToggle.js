import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from '../Dropdown';
import { Icon } from '../Icon';

// Must be a class component react-bootstrap passes a ref to this.
// eslint-disable-next-line react/prefer-stateless-function
class InfoTipToggle extends React.Component {
  static propTypes = {
    ...Dropdown.propTypes,
    children: PropTypes.node,
    open: PropTypes.bool,
    className: PropTypes.string
  };

  static defaultProps = {
    bsRole: 'toggle', // eslint-disable-line react/default-props-match-prop-types
    children: null,
    open: false,
    className: ''
  };

  render() {
    const { bsClass, bsRole, children, open, ...props } = this.props;

    return (
      <a href="#" aria-expanded={open} {...props}>
        <Icon type="pf" name="info" /> {children}
      </a>
    );
  }
}

export default InfoTipToggle;
