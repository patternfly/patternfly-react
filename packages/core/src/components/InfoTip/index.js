import { default as InfoTip } from './InfoTip';
import { default as InfoTipToggle } from './InfoTipToggle';
import { default as InfoTipMenu } from './InfoTipMenu';
import { default as InfoTipMenuFooter } from './InfoTipMenuFooter';
import { default as InfoTipMenuItemIcon } from './InfoTipMenuItemIcon';

InfoTip.Toggle = InfoTipToggle;
InfoTip.Menu = InfoTipMenu;
InfoTip.MenuItemIcon = InfoTipMenuItemIcon;
InfoTip.MenuFooter = InfoTipMenuFooter;

export {
  InfoTip,
  InfoTipToggle,
  InfoTipMenu,
  InfoTipMenuItemIcon,
  InfoTipMenuFooter
};
