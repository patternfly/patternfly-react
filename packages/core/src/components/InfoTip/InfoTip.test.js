import React from 'react';
import { shallow } from 'enzyme';
import InfoTip from './InfoTip';
import { KEY_CODES } from '../../common/helpers';
import { Dropdown } from '../Dropdown';

const props = {
  id: 'InfoTip',
  children: 'children'
};

test('renders Dropdown', () => {
  const view = shallow(<InfoTip {...props} />);
  expect(view).toMatchSnapshot('dropdown is rendered');
});

test('removes onToggle from passed props', () => {
  const onToggle = jest.fn();
  const view = shallow(<InfoTip {...props} onToggle={onToggle} />);
  expect(getDropdown(view).props().onToggle).not.toBe(onToggle);
});

test('toggles open and closed when keydown is enter', () => {
  const event = getMockKeyEvent(KEY_CODES.ENTER_KEY);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', event);
  expect(getDropdown(view).props().open).toBe(true);
  getDropdown(view).simulate('keydown', event);
  expect(getDropdown(view).props().open).toBe(false);
  expect(event.preventDefault).toHaveBeenCalledTimes(2);
});

test('tab keydown focuses footer and closes if footer is focused', () => {
  const tabKeyEvent = getMockKeyEvent(KEY_CODES.TAB_KEY);
  const enterKeyEvent = getMockKeyEvent(KEY_CODES.ENTER_KEY);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', enterKeyEvent); // open dropdown
  getDropdown(view).simulate('keydown', tabKeyEvent); // focus footer
  expect(getDropdown(view).props().open).toBe(true);
  getDropdown(view).simulate('keydown', tabKeyEvent); // close and unfocus footer
  expect(getDropdown(view).props().open).toBe(false);
});

test('tab keydown calls stopPropagation and stopImmediatePropagation', () => {
  const event = getMockKeyEvent(KEY_CODES.TAB_KEY);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', event); // focus footer
  expect(event.stopPropagation).toBeCalled();
  expect(event.nativeEvent.stopImmediatePropagation).toBeCalled();
});

test('closes on escape keydown', () => {
  const escKeyEvent = getMockKeyEvent(KEY_CODES.ESCAPE_KEY);
  const enterKeyEvent = getMockKeyEvent(KEY_CODES.ENTER_KEY);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', enterKeyEvent); // open dropdown
  getDropdown(view).simulate('keydown', escKeyEvent);
  expect(getDropdown(view).props().open).toBe(false);
});

test('closes if keydown includes shift key and a keycode', () => {
  const shiftKeyEvent = getMockKeyEvent(1, true);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', getMockKeyEvent(KEY_CODES.ENTER_KEY)); // open dropdown
  getDropdown(view).simulate('keydown', shiftKeyEvent);
  expect(getDropdown(view).props().open).toBe(false);
  getDropdown(view).simulate('keydown', shiftKeyEvent);
  expect(getDropdown(view).props().open).toBe(false);
});

test('does not close if keydown includes shift key and does not have a keycode', () => {
  const shiftKeyEvent = getMockKeyEvent(null, true);
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('keydown', getMockKeyEvent(KEY_CODES.ENTER_KEY)); // open dropdown
  getDropdown(view).simulate('keydown', shiftKeyEvent);
  expect(getDropdown(view).props().open).toBe(true);
});

test('onClick toggles Dropdown open', () => {
  const preventDefault = jest.fn();
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('click', { preventDefault });
  expect(getDropdown(view).props().open).toBe(true);
  getDropdown(view).simulate('click', { preventDefault });
  expect(getDropdown(view).props().open).toBe(false);
  expect(preventDefault).toHaveBeenCalledTimes(2);
});

test('onBlur closes Dropdown', () => {
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('click', { preventDefault: jest.fn() }); // open dropdown
  getDropdown(view).simulate('blur');
  expect(getDropdown(view).props().open).toBe(false);
});

test(`onBlur clicks related target`, () => {
  const relatedTargetClick = jest.fn();
  const view = shallow(<InfoTip {...props} />);
  getDropdown(view).simulate('blur', {
    relatedTarget: { click: relatedTargetClick }
  });
  expect(relatedTargetClick).toBeCalled();
});

function getDropdown(view) {
  return view.find(Dropdown);
}

function getMockKeyEvent(keyCode, shiftKey = false) {
  return {
    keyCode,
    shiftKey,
    preventDefault: jest.fn(),
    stopPropagation: jest.fn(),
    nativeEvent: {
      stopImmediatePropagation: jest.fn()
    }
  };
}
