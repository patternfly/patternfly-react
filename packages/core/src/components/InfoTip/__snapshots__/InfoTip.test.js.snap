// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dropdown: dropdown is rendered 1`] = `
<Uncontrolled(Dropdown)
  componentClass="li"
  id="InfoTip"
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onToggle={[Function]}
  open={false}
>
  children
</Uncontrolled(Dropdown)>
`;
